pkg/planner/core/partition_prune.go

```diff
@@ -38,6 +38,53 @@ func PartitionPruning(ctx base.PlanContext, tbl table.PartitionedTable, conds []
 			return nil, err
 		}
 		ret := s.convertToIntSlice(rangeOr, pi, partitionNames)
+		if pi.CanHaveOverlappingDroppingPartition() {
+			if len(ret) == 1 && ret[0] == FullRange {
+				ret = make([]int, 0, len(pi.Definitions))
+				for i := range pi.Definitions {
+					ret = append(ret, i)
+				}
+			}
+			newRet := make([]int, 0, len(ret))
+			for i := range ret {
+				idx := pi.GetOverlappingDroppingPartitionIdx(ret[i])
+				if idx == -1 {
+					continue
+				}
+				if idx == ret[i] {
+					newRet = append(newRet, idx)
+					continue
+				}
+				// partition being dropped, remove the consecutive range of dropping partitions
+				// and add the overlapping partition.
+				end := i + 1
+				for ; end < len(ret); end++ {
+					if ret[end] < idx {
+						continue
+					}
+					if ret[end] == idx {
+						// overlapping partition already included, append it together with the following
+						newRet = append(newRet, ret[end:]...)
+						if len(newRet) == len(pi.Definitions) {
+							return []int{FullRange}, nil
+						}
+						return newRet, nil
+					}
+					break
+				}
+				if len(partitionNames) == 0 || s.findByName(partitionNames, pi.Definitions[idx].Name.L) {
+					newRet = append(newRet, idx)
+				}
```

Comment:
When will this happen?

Reply:
This is when the overlapping partition was not included in the pruning result, so it should be added, unless explicit partition pruning is given in partitionNames, and it is not included there.
I have reformatted the code a little, with your suggestion above, and added comments to make it more clear.